Fixes #14771 - Show corresponding os list for arch

xprazak2 opened the pull request.

Commit:
Fixes #14771 - Show corresponding os list for arch

orrabin commented:
@xprazak2 I tested this and what I saw is that now  the list of operating systems fits the architecture but the architecture still doesn't change after you select the new hostgroup.

This does solve the description in the bug but another solution could be finding why the architecture doesn't change (this might be intentional but may be worth investigating).
Because this is a new host and os, media and partition table disappear when selecting a different hostgroup, I'd expect the architecture to also be removed or change.

xprazak2 commented:
@orrabin, you are right, the os, partition table and medium disappears when switching to a different hostgroup, but that turns out to be a related bug simulating expected behaviour. The host still has  values assigned from previous hostgroup even if they do not appear in the form.

Cause: hostgroup tries to provide [default values](https://github.com/theforeman/foreman/blob/develop/app/controllers/hosts_controller.rb#L636) for host. A value from hostgroup is not assigned if host already has something assigned, but parts of the form use [values from hostgroup](https://github.com/theforeman/foreman/blob/develop/app/controllers/hosts_controller.rb#L620-L626) regardless. This leads to discrepancy between what can be seen in the form and what host has assigned.

For now, I would just try to make the form consistent and display the values that host actually has assigned, but providing defaults for a new host in this manner seems a bit confusing, especially when puppet classes change with each change of a hostgroup.

xprazak2 commented:
The form should be consistent and show the actual configuration of a host. I did not change the current behaviour - hostgroup provides defaults if there are none.

xprazak2 commented:
[test]

xprazak2 commented:
When hostgroup for a new host is changed, fields in Operating System tab are updated with values from newly selected hostgroup. Behaviour for existing hosts remains unchanged.

dLobatog commented on test/active_support_test_case_helper.rb:
```diff
@@ -56,7 +56,7 @@ def reset_setting_cache
   alias_method :assert_not_include, :assert_not_includes
   class <<self
-    alias_method :test, :it
+    # alias_method :test, :it
```
I'm almost certain this is going to make tests fail, sorry

dLobatog replied:
I ran them locally, and I'm surprised to see this pass https://github.com/theforeman/foreman/blob/develop/test/functional/application_controller_subclass_test.rb All good then (remove the line instead of commenting it)

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -961,10 +961,14 @@ def assign_hostgroup_attributes(attrs = [])
   end
 
+  def assign_hostgroup_attribute(attr, value)
+    self.send("#{attr}=", value) if new_record? || !send(attr).present?
```
Nitpick, does `new_record?` do anything here? For a new record, `send(attr).present?` is always going to be false isn't it?

dLobatog replied:
Ah I see it now, you want to preserve the options  for existing hosts. Can you move this check to where this method is called? `assign_hostgroup_attribute(attr, value) if new_record?`

dLobatog commented:
@xprazak2 Thanks, it seems to solve the problem. Just made a comment about making  `assign_hostgroup_attribute` and removing the `alias_method :test, :it` line, I'll merge if Jenkins is :green_apple:  after that.

dLobatog commented:
[test] job was aborted

xprazak2 commented:
I moved the `new_record?` check to `set_hostgroup_defaults` where `assign_hostgroup_attribute` is called. This slightly changes behaviour for existing hosts - it used to preserve non-empty attributes, now it preserves all of them.

I also removed the alias from `active_support_test_case_helper`.

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -961,7 +962,7 @@ def assign_hostgroup_attributes(attrs = [])
       next if send(attr).to_i == -1
       value = hostgroup.send("inherited_#{attr}")
-      self.send("#{attr}=", value) unless send(attr).present?
+      self.send("#{attr}=", value)
```
I think the test is failing because of this line lacking `unless send(attr).present?` - this preserves the behavior we had, hostgroup selection overrides only empty values.

dLobatog commented:
@xprazak2 Thanks, remember to ping me after rebasing, I would've merged this long ago if I knew it was ready!

dLobatog commented:
Merged as d301d11d8141d924f20c4b6136f1607f8052b154, thanks @xprazak2!
